dialogue-clients/src/main/java/com/palantir/dialogue/clients/ReloadingClientFactory.java

```diff
@@ -239,46 +243,75 @@ public PerHostClientFactory perHost(String serviceName) {
+                        selfAddresses = Set.of();
+                    }
+
+                    ImmutableMap.Builder<PerHostTarget, Channel> map = ImmutableMap.builder();
                     for (int i = 0; i < targetUris.size(); i++) {
                         TargetUri targetUri = targetUris.get(i);
                         ServiceConfiguration singleUriServiceConf = ServiceConfiguration.builder()
                                 .from(serviceConfiguration)
                                 .uris(ImmutableList.of(targetUri.uri()))
                                 .build();
+                        boolean isSelf = targetUri
+                                .resolvedAddress()
+                                .map(selfAddresses::contains)
+                                .orElse(false);
 
                         // subtle gotcha here is that every single one of these has the same channelName,
                         // which means metrics like the QueuedChannel counter will end up being the sum of all of them.
-                        list.add(cache.getNonReloadingChannel(
-                                params,
-                                singleUriServiceConf,
-                                ImmutableList.of(targetUri),
-                                channelName,
-                                OptionalInt.of(i)));
+                        map.put(
+                                new PerHostTarget(targetUri, isSelf),
+                                cache.getNonReloadingChannel(
+                                        params,
+                                        singleUriServiceConf,
+                                        ImmutableList.of(targetUri),
+                                        channelName,
+                                        OptionalInt.of(i)));
                     }
-                    return list.build();
+                    return map.buildKeepingLast();
                 });
 
         return new PerHostClientFactory() {
             @Override
             public Refreshable<List<Channel>> getPerHostChannels() {
-                return perHostDialogueChannels.map(ImmutableList::copyOf);
+                return perHostChannels.map(channels -> ImmutableList.copyOf(channels.values()));
+            }
+
+            @Override
+            public Refreshable<Map<PerHostTarget, Channel>> getNamedPerHostChannels() {
+                return perHostChannels;
             }
 
             @Override
             public <T> Refreshable<List<T>> getPerHost(Class<T> clientInterface) {
-                return getPerHostChannels().map(channels -> channels.stream()
-                        .map(chan -> Reflection.callStaticFactoryMethod(clientInterface, chan, params.runtime()))
+                return perHostChannels.map(channels -> channels.values().stream()
+                        .map(channel -> Reflection.callStaticFactoryMethod(clientInterface, channel, params.runtime()))
```

Comment:
Thoughts on delegation here?
```java
return getNamedPerHost(clientInterface).map(result -> result.values().asList());
```

Reply:
Should we make the existing methods default methods on `PerHostClientFactory`? Or would you prefer to keep the implementation here?

Reply:
Delegation would require me to change the return type declared in `PerHostClientFactory` to be `ImmutableMap`, which probably isn't what we want to do here.

Reply:
> Delegation would require me to change the return type declared in PerHostClientFactory to be ImmutableMap, which probably isn't what we want to do here.

Ah, we didn't handle PECS (Producer Extends Consumer Super) very well on the `PerHostClientFactory` existing interface methods -- ideally we may have declared something like `Refreshable<? extends List<? extends Channel>> getPerHostChannels();` for greater flexibility in the implementation, however that does force callers who don't use the `var` keyword to declare a fair bit more cruft.

Happy to leave this as is